src/main/java/org/betonquest/betonquest/api/profiles/ProfileValueMap.java

```diff
@@ -0,0 +1,149 @@
+     */
+    public ProfileValueMap(final Server server, final Class<T> clazz) {
+        this.profileMap = new HashMap<>();
+        this.type = clazz;
+        this.server = server;
+    }
+
+    @Override
+    public int size() {
+        return profileMap.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return profileMap.isEmpty();
+    }
+
+    @Override
+    public boolean containsKey(final Object key) {
+        if (type.isInstance(key)) {
+            return profileMap.containsKey(key);
+        }
+        return false;
+    }
+
+    @Override
+    public boolean containsValue(final Object value) {
+        if (value instanceof Profile profile) {
+            return profileMap.containsValue(profile.getProfileUUID());
+        }
+        return false;
+    }
+
+    @Override
+    public Profile get(final Object key) {
+        if (type.isInstance(key)) {
+            return PlayerConverter.getID(server.getPlayer(profileMap.get(key)));
+        }
+        return null;
+    }
+
+    @Nullable
+    @Override
+    public Profile put(final T key, final Profile value) {
+        profileMap.put(key, value.getProfileUUID());
+        return value;
+    }
+
+    @Override
+    public Profile remove(final Object key) {
+        if (type.isInstance(key)) {
+            return PlayerConverter.getID(server.getPlayer(profileMap.remove(key)));
+        }
+        return null;
+    }
+
+    @Override
+    public void putAll(@NotNull final Map<? extends T, ? extends Profile> map) {
+        final Map<T, UUID> entries = map.entrySet().stream()
+                .collect(Collectors.toMap(Entry::getKey, entry -> entry.getValue().getProfileUUID()));
+        profileMap.putAll(entries);
+    }
+
+    @Override
+    public void clear() {
+        profileMap.clear();
+    }
+
+    @NotNull
+    @Override
+    public Set<T> keySet() {
+        return profileMap.keySet();
+    }
+
+    @NotNull
+    @Override
+    public Collection<Profile> values() {
+        return profileMap.values().stream()
+                .map(uuid -> PlayerConverter.getID(server.getPlayer(uuid)))
+                .collect(Collectors.toList());
+    }
+
+    @NotNull
+    @Override
+    public Set<Entry<T, Profile>> entrySet() {
+        return profileMap.entrySet().stream()
+                .map(entry -> new Entry<T, Profile>() {
+                    @Override
+                    public T getKey() {
+                        return entry.getKey();
+                    }
+
+                    @Override
+                    public Profile getValue() {
+                        return PlayerConverter.getID(server.getPlayer(entry.getValue()));
+                    }
+
+                    @Override
+                    public Profile setValue(final Profile value) {
+                        entry.setValue(value.getProfileUUID());
+                        return value;
+                    }
+                })
+                .collect(Collectors.toSet());
```

Comment:
This implementation violates the interface definition: https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/util/Map.html#entrySet()

The returned `Set` must be linked to the map, so `remove()` calls on the map (or it's iterator) must also remove the entries from this map.